File: Source/Client/Game/Sound.vb
Imports System.IO
Imports System.Xml
Imports Core
Imports ManagedBass
Imports ManagedBass.Midi

Module Sound

    ' Sound and Music handles for ManagedBass
    Friend MusicStream As Integer
    Friend SoundStream As Integer
    Friend WeatherStream As Integer
    Friend ExtraSoundStream As Integer

    Friend MusicCache() As String
    Friend SoundCache() As String

    Friend FadeInSwitch As Boolean
    Friend FadeOutSwitch As Boolean
    Friend CurrentMusic As String
    Friend CurrentWeatherMusic As String

    Friend SoundFontHandle As Integer

    Sub PlayMusic(fileName As String)
        Dim path = IO.Path.Combine(Core.Path.Music, fileName)

        If fileName = "None" Then
            Exit Sub
        End If

        If Core.Settings.Music = 0 Or Not File.Exists(path) Then
            StopMusic()
            Exit Sub
        End If

        If fileName = CurrentMusic Then
            Exit Sub
        End If

        If IO.Path.GetExtension(fileName).ToLower() = ".mid" Then
            StopMusic()
            PlayMidi(path)
            CurrentMusic = fileName
            Exit Sub
        End If

        Try
            StopMusic() ' Stop any currently playing music before starting a new one

            MusicStream = Bass.CreateStream(path, 0, 0, BassFlags.Loop)
            If MusicStream <> 0 Then
                Bass.ChannelPlay(MusicStream)
                Bass.ChannelSetAttribute(MusicStream, ChannelAttribute.Volume, Core.Settings.MusicVolume / 100.0F)
                CurrentMusic = fileName
                FadeInSwitch = 1
            End If
        Catch ex As Exception
            Console.WriteLine($"Error playing music: {ex.Message}")
        End Try
    End Sub

    Sub InitializeBASS()
        ' Initialize BASS with the default output device
        If Not Bass.Init(-1, 44100, DeviceInitFlags.Default) Then
            Console.WriteLine($"Failed to initialize BASS. Error: {Bass.LastError}")
            Exit Sub
        End If

        ' Load the SoundFont (.sf2) for MIDI playback
        Dim soundFontPath As String = "GeneralUser.sf2"
        If Not File.Exists(soundFontPath) Then
            Console.WriteLine($"SoundFont not found: {soundFontPath}")
            Exit Sub
        End If

        ' Initialize the SoundFont
        SoundFontHandle = BassMidi.FontInit(soundFontPath, BassFlags.Default)
        If SoundFontHandle = 0 Then
            Console.WriteLine($"Failed to load SoundFont. Error: {Bass.LastError}")
            Exit Sub
        End If

        ' Set the volume for the SoundFont
        BassMidi.FontSetVolume(SoundFontHandle, 1.0F) ' 100% volume

        CacheMusic()
        CacheSound()
    End Sub

    Sub PlayMidi(filePath As String)
        StopMusic() ' Ensure previous music is stopped

        ' Load and play the MIDI file using ManagedBass.Midi
        MusicStream = BassMidi.CreateStream(filePath, 0, 0, BassFlags.Loop, 44100)

        ' Correctly set the SoundFont for the MIDI stream
        Dim font As New MidiFont With {
            .Handle = SoundFontHandle, ' Use the SoundFont handle
            .Preset = -1, ' -1 means all presets from the SoundFont
            .Bank = 0 ' Bank 0 (General MIDI bank)
        }

        ' Create an array with the MidiFont structure
        Dim fonts() As MidiFont = {font}

        ' Set the fonts for the MIDI stream and check if it succeeded
        Dim fontCount As Integer = BassMidi.StreamSetFonts(MusicStream, fonts, fonts.Length)

        If fontCount = 0 Then
            Console.WriteLine($"Failed to assign SoundFont. Error: {Bass.LastError}")
        End If

        ' Ensure the file exists before attempting to load it
        If Not File.Exists(filePath) Then
            Console.WriteLine($"MIDI file not found: {filePath}")
            Exit Sub
        End If

        If MusicStream <> 0 Then
            Bass.ChannelPlay(MusicStream)
            Bass.ChannelSetAttribute(MusicStream, ChannelAttribute.Volume, Core.Settings.MusicVolume / 100.0F)
        Else
            ' Log the last error if stream creation fails
            Dim errorCode As Errors = Bass.LastError
            Console.WriteLine($"Failed to load MIDI file. Error: {errorCode}")
        End If
    End Sub

    Sub StopMusic()
        If MusicStream <> 0 Then
            Bass.ChannelStop(MusicStream)
            Bass.StreamFree(MusicStream)
            MusicStream = 0
            CurrentMusic = ""
        End If
    End Sub

    Sub PlaySound(fileName As String, x As Integer, y As Integer, Optional looped As Boolean = False)
        If Core.Settings.Sound = 0 Or Not File.Exists(Core.Path.Sounds & fileName) Then Exit Sub

        Try
            StopSound() ' Stop previous sound if any

            SoundStream = Bass.CreateStream(Core.Path.Sounds & fileName, 0, 0, If(looped, BassFlags.Loop, BassFlags.Default))
            If SoundStream <> 0 Then
                Dim calculatedVolume As Double = CalculateSoundVolume(x, y)
                Bass.ChannelSetAttribute(SoundStream, ChannelAttribute.Volume, calculatedVolume / 100.0F)
                Bass.ChannelPlay(SoundStream)
            End If
        Catch ex As Exception
            Console.WriteLine($"Failed to load sound: {ex.Message}")
        End Try
    End Sub

    Sub StopSound()
        If SoundStream <> 0 Then
            Bass.ChannelStop(SoundStream)
            Bass.StreamFree(SoundStream)
            SoundStream = 0
        End If
    End Sub

    Sub PlayExtraSound(fileName As String, Optional looped As Boolean = False)
        If Core.Settings.Sound = 0 Or Not File.Exists(Core.Path.Sounds & fileName) Then Exit Sub

        Try
            StopExtraSound()

            ExtraSoundStream = Bass.CreateStream(Core.Path.Sounds & fileName, 0, 0, If(looped, BassFlags.Loop, BassFlags.Default))
            If ExtraSoundStream <> 0 Then
                Bass.ChannelSetAttribute(ExtraSoundStream, ChannelAttribute.Volume, Core.Settings.SoundVolume / 100.0F)
                Bass.ChannelPlay(ExtraSoundStream)
            End If
        Catch ex As Exception
            Console.WriteLine($"Failed to load extra sound: {ex.Message}")
        End Try
    End Sub

    Sub StopExtraSound()
        If ExtraSoundStream <> 0 Then
            Bass.ChannelStop(ExtraSoundStream)
            Bass.StreamFree(ExtraSoundStream)
            ExtraSoundStream = 0
        End If
    End Sub

    ' The fade methods should directly adjust the volume of the ManagedBass streams
    Sub FadeIn()
        If MusicStream <> 0 Then
            Dim currentVolume As Single
            Bass.ChannelGetAttribute(MusicStream, ChannelAttribute.Volume, currentVolume)
            If currentVolume < Core.Settings.MusicVolume / 100.0F Then
                Bass.ChannelSetAttribute(MusicStream, ChannelAttribute.Volume, currentVolume + 0.03F)
            Else
                FadeInSwitch = 0
            End If
        End If
    End Sub

    Sub FadeOut()
        If MusicStream <> 0 Then
            Dim currentVolume As Single
            Bass.ChannelGetAttribute(MusicStream, ChannelAttribute.Volume, currentVolume)
            If currentVolume > 0 Then
                Bass.ChannelSetAttribute(MusicStream, ChannelAttribute.Volume, currentVolume - 0.03F)
            Else
                FadeOutSwitch = 0
                StopMusic()
            End If
        End If
    End Sub

    Public Function CalculateSoundVolume(ByRef x As Integer, ByRef y As Integer) As Double
        Dim X1, X2, Y1, Y2 As Integer
        Dim Distance As Double

        If Not GameState.InGame = True Then
            CalculateSoundVolume = 1
            Return CalculateSoundVolume
        End If

        If GameState.InGame AndAlso x = GetPlayerX(State.MyIndex) AndAlso y = GetPlayerY(State.MyIndex) Then
            CalculateSoundVolume = 1
            CalculateSoundVolume *= Core.Settings.SoundVolume
            Return CalculateSoundVolume
        End If

        If x > -1 OrElse y > -1 Then
            If x = -1 Then x = 0
            If y = -1 Then y = 0
            X1 = CInt((Core.Type.Player(State.MyIndex).X * 32) + Core.Type.Player(State.MyIndex).XOffset)
            Y1 = CInt((Core.Type.Player(State.MyIndex).Y * 32) + Core.Type.Player(State.MyIndex).YOffset)
            X2 = x * 32
            Y2 = y * 32

            If CInt(Math.Pow((X2 - X1), 2)) + CInt(Math.Pow((Y2 - Y1), 2)) < 0 Then
                Distance = Math.Sqrt(CInt(Math.Pow((X2 - X1), 2)) + CInt(Math.Pow((Y2 - Y1), 2)) * -1)
            Else
                Distance = Math.Sqrt(CInt(Math.Pow((X2 - X1), 2)) + CInt(Math.Pow((Y2 - Y1), 2)))
            End If

            ' If the range is greater than 32, do not send a sound
            If (Distance / 32) > 32 Then
                CalculateSoundVolume = 0
            Else
                CalculateSoundVolume = 1 / (Distance / 32)

                If CalculateSoundVolume > 1 Then
                    CalculateSoundVolume = 1
                ElseIf CalculateSoundVolume < 0 Then
                    CalculateSoundVolume *= -1
                End If
            End If
        Else
            CalculateSoundVolume = 1
        End If

        CalculateSoundVolume *= Core.Settings.SoundVolume

        Return CalculateSoundVolume
    End Function

    Sub FreeBASS()
        If SoundFontHandle <> 0 Then
            BassMidi.FontFree(SoundFontHandle)
        End If
        Bass.Free()
    End Sub

    Sub PlayWeatherSound(fileName As String, Optional looped As Boolean = False)
        ' Check if sound is enabled and the file exists
        If Not Core.Settings.Sound = 1 Or Not File.Exists(Core.Path.Sounds & fileName) Then Exit Sub

        ' Avoid reloading the same sound if it's already playing
        If CurrentWeatherMusic = fileName Then Exit Sub

        ' Stop any previously playing weather sound
        StopWeatherSound()

        ' Load the new sound file
        Dim soundPath As String = Core.Path.Sounds & fileName
        WeatherStream = Bass.CreateStream(soundPath, 0, 0, If(looped, BassFlags.Loop, BassFlags.Default))

        ' Check if the stream was created successfully
        If WeatherStream <> 0 Then
            ' Play the sound
            Bass.ChannelSetAttribute(WeatherStream, ChannelAttribute.Volume, Core.Settings.SoundVolume / 100.0F)
            Bass.ChannelPlay(WeatherStream, False)
            CurrentWeatherMusic = fileName
        Else
            Console.WriteLine("Failed to play sound: " & Bass.LastError.ToString())
        End If
    End Sub

    Sub StopWeatherSound()
        ' Stop and free the current weather sound channel
        If WeatherStream <> 0 Then
            Bass.ChannelStop(WeatherStream)
            Bass.StreamFree(WeatherStream)
            WeatherStream = 0
        End If

        CurrentWeatherMusic = ""
    End Sub

End Module

File: Source/Client/Game/Sound.cs
using System;
using System.IO;
using Core;
using static Core.Global.Command;
using ManagedBass;
using ManagedBass.Midi;
using Microsoft.VisualBasic.CompilerServices;

namespace Client
{

    public class Sound
    {

        // Sound and Music handles for ManagedBass
        public static int MusicStream;
        public static int SoundStream;
        public static int WeatherStream;
        public static int ExtraSoundStream;

        public static string[] MusicCache;
        public static string[] SoundCache;

        public static bool FadeInSwitch;
        public static bool FadeOutSwitch;
        public static string CurrentMusic;
        public static string CurrentWeatherMusic;

        public static int SoundFontHandle;

        public static void PlayMusic(string fileName)
        {
            string path = System.IO.Path.Combine(Core.Path.Music, fileName);

            if (fileName == "None")
            {
                return;
            }

            if (Conversions.ToInteger(SettingsManager.Instance.Music) == 0 | !File.Exists(path))
            {
                StopMusic();
                return;
            }

            if ((fileName ?? "") == (CurrentMusic ?? ""))
            {
                return;
            }

            if (System.IO.Path.GetExtension(fileName).ToLower() == ".mid")
            {
                StopMusic();
                PlayMidi(path);
                CurrentMusic = fileName;
                return;
            }

            try
            {
                StopMusic(); // Stop any currently playing music before starting a new one

                MusicStream = Bass.CreateStream(path, 0L, 0L, BassFlags.Loop);
                if (MusicStream != 0)
                {
                    Bass.ChannelPlay(MusicStream);
                    Bass.ChannelSetAttribute(MusicStream, ChannelAttribute.Volume, SettingsManager.Instance.MusicVolume / 100.0f);
                    CurrentMusic = fileName;
                    FadeInSwitch = true;
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error playing music: {ex.Message}");
            }
        }

        public static void InitializeBASS()
        {
            // Initialize BASS with the default output device
            if (!Bass.Init(-1, 44100, DeviceInitFlags.Default))
            {
                Console.WriteLine($"Failed to initialize BASS. Error: {Bass.LastError}");
                return;
            }

            // Load the SoundFont (.sf2) for MIDi playback
            string soundFontPath = "GeneralUser.sf2";
            if (!File.Exists(soundFontPath))
            {
                Console.WriteLine($"SoundFont not found: {soundFontPath}");
                return;
            }

            // Initialize the SoundFont
            SoundFontHandle = BassMidi.FontInit(soundFontPath, (FontInitFlags)BassFlags.Default);
            if (SoundFontHandle == 0)
            {
                Console.WriteLine($"Failed to load SoundFont. Error: {Bass.LastError}");
                return;
            }

            // Set the volume for the SoundFont
            BassMidi.FontSetVolume(SoundFontHandle, 1.0f); // 100% volume

            General.CacheMusic();
            General.CacheSound();
        }

        public static void PlayMidi(string filePath)
        {
            StopMusic(); // Ensure previous music is stopped

            // Load and play the MIDi file using ManagedBass.Midi
            MusicStream = BassMidi.CreateStream(filePath, 0L, 0L, BassFlags.Loop, 44100);

            // Correctly set the SoundFont for the MIDi stream
            var font = new MidiFont()
            {
                Handle = SoundFontHandle, // Use the SoundFont handle
                Preset = -1, // -1 means all presets from the SoundFont
                Bank = 0 // Bank 0 (General MIDi bank)
            };

            // Create an array with the MidiFont structure
            MidiFont[] fonts = new MidiFont[] { font };

            // Set the fonts for the MIDi stream and check if it succeeded
            int fontCount = BassMidi.StreamSetFonts(MusicStream, fonts, fonts.Length);

            if (fontCount == 0)
            {
                Console.WriteLine($"Failed to assign SoundFont. Error: {Bass.LastError}");
            }

            // Ensure the file exists before attempting to load it
            if (!File.Exists(filePath))
            {
                Console.WriteLine($"MIDi file not found: {filePath}");
                return;
            }

            if (MusicStream != 0)
            {
                Bass.ChannelPlay(MusicStream);
                Bass.ChannelSetAttribute(MusicStream, ChannelAttribute.Volume, SettingsManager.Instance.MusicVolume / 100.0f);
            }
            else
            {
                // Log the last error if stream creation fails
                var errorCode = Bass.LastError;
                Console.WriteLine($"Failed to load MIDi file. Error: {errorCode}");
            }
        }

        public static void StopMusic()
        {
            if (MusicStream != 0)
            {
                Bass.ChannelStop(MusicStream);
                Bass.StreamFree(MusicStream);
                MusicStream = 0;
                CurrentMusic = "";
            }
        }

        public static void PlaySound(string fileName, int x, int y, bool looped = false)
        {
            if (Conversions.ToInteger(SettingsManager.Instance.Sound) == 0 | !File.Exists(Core.Path.Sounds + fileName))
                return;

            try
            {
                StopSound(); // Stop previous sound if any

                SoundStream = Bass.CreateStream(Core.Path.Sounds + fileName, 0L, 0L, looped ? BassFlags.Loop : BassFlags.Default);
                if (SoundStream != 0)
                {
                    double calculatedVolume = CalculateSoundVolume(ref x, ref y);
                    Bass.ChannelSetAttribute(SoundStream, ChannelAttribute.Volume, calculatedVolume / 100.0d);
                    Bass.ChannelPlay(SoundStream);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Failed to load sound: {ex.Message}");
            }
        }

        public static void StopSound()
        {
            if (SoundStream != 0)
            {
                Bass.ChannelStop(SoundStream);
                Bass.StreamFree(SoundStream);
                SoundStream = 0;
            }
        }

        public static void PlayExtraSound(string fileName, bool looped = false)
        {
            if (Conversions.ToInteger(SettingsManager.Instance.Sound) == 0 | !File.Exists(Core.Path.Sounds + fileName))
                return;

            try
            {
                StopExtraSound();

                ExtraSoundStream = Bass.CreateStream(Core.Path.Sounds + fileName, 0L, 0L, looped ? BassFlags.Loop : BassFlags.Default);
                if (ExtraSoundStream != 0)
                {
                    Bass.ChannelSetAttribute(ExtraSoundStream, ChannelAttribute.Volume, SettingsManager.Instance.SoundVolume / 100.0f);
                    Bass.ChannelPlay(ExtraSoundStream);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Failed to load extra sound: {ex.Message}");
            }
        }

        public static void StopExtraSound()
        {
            if (ExtraSoundStream != 0)
            {
                Bass.ChannelStop(ExtraSoundStream);
                Bass.StreamFree(ExtraSoundStream);
                ExtraSoundStream = 0;
            }
        }

        // The fade methods should directly adjust the volume of the ManagedBass streams
        public static void FadeIn()
        {
            if (MusicStream != 0)
            {
                float currentVolume;
                Bass.ChannelGetAttribute(MusicStream, ChannelAttribute.Volume, out currentVolume);
                if (currentVolume < SettingsManager.Instance.MusicVolume / 100.0f)
                {
                    Bass.ChannelSetAttribute(MusicStream, ChannelAttribute.Volume, currentVolume + 0.03f);
                }
                else
                {
                    FadeInSwitch = false;
                }
            }
        }

        public static void FadeOut()
        {
            if (MusicStream != 0)
            {
                float currentVolume;
                Bass.ChannelGetAttribute(MusicStream, ChannelAttribute.Volume, out currentVolume);
                if (currentVolume > 0f)
                {
                    Bass.ChannelSetAttribute(MusicStream, ChannelAttribute.Volume, currentVolume - 0.03f);
                }
                else
                {
                    FadeOutSwitch = false;
                    StopMusic();
                }
            }
        }

        public static double CalculateSoundVolume(ref int x, ref int y)
        {
            double CalculateSoundVolumeRet = default;
            int x1, x2, y1, y2;
            double Distance;

            if (!(GameState.InGame == true))
            {
                CalculateSoundVolumeRet = 1d;
                return CalculateSoundVolumeRet;
            }

            if (GameState.InGame && x == GetPlayerX(GameState.MyIndex) && y == GetPlayerY(GameState.MyIndex))
            {
                CalculateSoundVolumeRet = 1d;
                CalculateSoundVolumeRet *= SettingsManager.Instance.SoundVolume;
                return CalculateSoundVolumeRet;
            }

            if (x > -1 || y > -1)
            {
                if (x == -1)
                    x = 0;
                if (y == -1)
                    y = 0;
                x1 = Core.Data.Player[GameState.MyIndex].X;
                y1 = Core.Data.Player[GameState.MyIndex].Y;
                x2 = x * 32;
                y2 = y * 32;

                if ((int)Math.Round(Math.Pow(x2 - x1, 2d)) + (int)Math.Round(Math.Pow(y2 - y1, 2d)) < 0)
                {
                    Distance = Math.Sqrt((int)Math.Round(Math.Pow(x2 - x1, 2d)) + (int)Math.Round(Math.Pow(y2 - y1, 2d)) * -1);
                }
                else
                {
                    Distance = Math.Sqrt((int)Math.Round(Math.Pow(x2 - x1, 2d)) + (int)Math.Round(Math.Pow(y2 - y1, 2d)));
                }

                // If the range is greater than 32, do not send a sound
                if (Distance / 32d > 32d)
                {
                    CalculateSoundVolumeRet = 0d;
                }
                else
                {
                    CalculateSoundVolumeRet = 1d / (Distance / 32d);

                    if (CalculateSoundVolumeRet > 1d)
                    {
                        CalculateSoundVolumeRet = 1d;
                    }
                    else if (CalculateSoundVolumeRet < 0d)
                    {
                        CalculateSoundVolumeRet *= -1;
                    }
                }
            }
            else
            {
                CalculateSoundVolumeRet = 1d;
            }

            CalculateSoundVolumeRet *= SettingsManager.Instance.SoundVolume;

            return CalculateSoundVolumeRet;
        }

        public static void FreeBASS()
        {
            if (SoundFontHandle != 0)
            {
                BassMidi.FontFree(SoundFontHandle);
            }
            Bass.Free();
        }

        public static void PlayWeatherSound(string fileName, bool looped = false)
        {
            // Check if sound is enabled and the file exists
            if (!(Conversions.ToInteger(SettingsManager.Instance.Sound) == 1) | !File.Exists(Core.Path.Sounds + fileName))
                return;

            // Avoid reloading the same sound if it's already playing
            if ((CurrentWeatherMusic ?? "") == (fileName ?? ""))
                return;

            // Stop any previously playing weather sound
            StopWeatherSound();

            // Load the new sound file
            string soundPath = Core.Path.Sounds + fileName;
            WeatherStream = Bass.CreateStream(soundPath, 0L, 0L, looped ? BassFlags.Loop : BassFlags.Default);

            // Check if the stream was created successfully
            if (WeatherStream != 0)
            {
                // Play the sound
                Bass.ChannelSetAttribute(WeatherStream, ChannelAttribute.Volume, SettingsManager.Instance.SoundVolume / 100.0f);
                Bass.ChannelPlay(WeatherStream, false);
                CurrentWeatherMusic = fileName;
            }
            else
            {
                Console.WriteLine("Failed to play sound: " + Bass.LastError.ToString());
            }
        }

        public static void StopWeatherSound()
        {
            // Stop and free the current weather sound channel
            if (WeatherStream != 0)
            {
                Bass.ChannelStop(WeatherStream);
                Bass.StreamFree(WeatherStream);
                WeatherStream = 0;
            }

            CurrentWeatherMusic = "";
        }

    }
}